bundler/helpers/v1/monkey_patches/lazy_specification_patch.rb

```diff
@@ -0,0 +1,18 @@
+# typed: false
+# frozen_string_literal: true
+
+# Adds default_gem? method to Bundler::LazySpecification
+# https://github.com/rubygems/rubygems/blob/bundler-v2.5.9/lib/rubygems/basic_specification.rb#L96-L99
+
+require "bundler/lazy_specification"
+
+# Check if the gem is a default gem
+# Ensure loaded_from is not nil and check if the gem is located in the default specifications directory
+module LazySpecificationDefaultGemPatch
+  def default_gem?
+    loaded_from && File.dirname(loaded_from) == Gem.default_specifications_dir
```

Comment:
I know we looked at this for bundler v2 but I just wanted to double check this for v1 and of course it turned out to be a bit of a rabbit hole. Here's my thought process looking at the source:
* [We use bundler 1.17.3](https://github.com/dependabot/dependabot-core/blob/29dcdfa8c7b91c011a2aefb9389a9ad025285000/bundler/helpers/v1/build#L25) to run v1 helpers.
* [`LazySpecification`](https://github.com/rubygems/bundler/blob/v1.17.3/lib/bundler/lazy_specification.rb) exists in bundler 1.17.3 but it does not define this `#loaded_from` method. However it is using [`#method_missing`](https://github.com/rubygems/bundler/blob/d7089abb6a3c922bfe74199c103e745bb11bd46f/lib/bundler/lazy_specification.rb#L115) to delegate to a wrapped specification, an `EndpointSpecification`, a `RemoteSpecification`, or a `StubSpecification`, by my read. `EndpointSpecification` inherits from `Gem::Specification`, the other two do not.
* The minimum version of rubygems [we require](https://github.com/dependabot/dependabot-core/blob/29dcdfa8c7b91c011a2aefb9389a9ad025285000/common/dependabot-common.gemspec#L24) is 3.3.7.
* In rubygems >= 3.3.7, `Gem::Specification` defines [`#loaded_from`](https://github.com/rubygems/rubygems/blob/bafe43c59332cd37d464e03b666233827e7704c3/lib/rubygems/basic_specification.rb#L25) (which appears to be [method chain-aliased](https://github.com/rubygems/rubygems/blob/d7089abb6a3c922bfe74199c103e745bb11bd46f/lib/bundler/rubygems_ext.rb#L41) in bundler 1.17.3 so it can define [its own `#loaded_from`](https://github.com/rubygems/rubygems/blob/d7089abb6a3c922bfe74199c103e745bb11bd46f/lib/bundler/rubygems_ext.rb#L57))

This is where I'm going to put a pin in it. I first tried loading up bundler 1.17.3 with ruby 3.3.1/rubygems 3.5.9 in the repl to poke around but I wasn't able to verify.

It would also be nice to figure out the specific version of bundler in which this monkey patch became necessary (`#default_gem?` exists on `Gem::Specification` in versions 3.3.7 through latest, so it's a bundler thing).

Reply:
As Abdul suggested, it would be good to verify this somehow using bundler v1.

Reply:
```suggestion
    respond_to?(:loaded_from) && loaded_from && File.dirname(loaded_from) == Gem.default_specifications_dir
```
Updated to suggest `respond_to?` to alleviate my concern about a possible `NoMethodError`. I would feel ok shipping with this change for both v1 and v2, with the understanding that this is only a temporary fix and we should really try to fix it upstream if possible and remove the monkey patches.

Reply:
@honeyankit sorry, do you mind making the same change as in b0fea74 for v2 as well? Just to be safe.

Reply:
@bdragon I will do that. I have not looked at this PR yet. Still trying to reproduce the issue

Reply:
I can't help but wonder if this is a case where dropping support for Bundler v1 would have saved a day+ of engineering time that we could put toward better support of Bundler v2...